File: src/utils/airtable.py
from pyairtable import Table
from datetime import datetime
import os
from dotenv import load_dotenv

load_dotenv()

class AirtableClient:
    def __init__(self):
        self.table = Table(
            os.getenv('AIRTABLE_API_KEY'),
            os.getenv('AIRTABLE_BASE_ID'),
            os.getenv('AIRTABLE_TABLE_NAME')
        )
    
    def get_user(self, telegram_id: str):
        """Get user by Telegram ID"""
        records = self.table.all(formula=f"{{telegram_id}}='{telegram_id}'")
        return records[0] if records else None
    
    def create_user(self, telegram_id: str, username: str):
        """Create new user"""
        return self.table.create({
            "telegram_id": telegram_id,
            "username": username,
            "status": "free"
        })
        
    def update_user_status(self, telegram_id: str, status: str):
        """Update user subscription status"""
        user = self.get_user(telegram_id)
        if user:
            self.table.update(user['id'], {
                'status': status,
                'subscription_updated': datetime.now().isoformat()
            })
            return True
        return False

    def get_subscription_status(self, telegram_id: str) -> str:
        """Get user's subscription status"""
        user = self.get_user(telegram_id)
        if user:
            return user['fields'].get('status', 'free')
        return 'free'
        
    def update_payment_details(self, telegram_id: str, details: dict):
        """Update payment details for user"""
        user = self.get_user(telegram_id)
        if user:
            self.table.update(user['id'], {
                'payment_date': details['payment_date'],
                'payment_amount': details['amount_paid'],
                'transaction_id': details['transaction_signature'],
                'payment_status': details['payment_status']
            })
            return True
        return False
        
    def get_all_users(self):
        """Get all users from Airtable"""
        try:
            return self.table.all()
        except Exception as e:
            print(f"Error fetching users from Airtable: {e}")
            return []
            
    async def createListing(self, listing_data):
        """Create a new listing in Airtable"""
        try:
            # Create a new table instance for listings
            listings_table = Table(
                os.getenv('AIRTABLE_API_KEY'),
                os.getenv('AIRTABLE_BASE_ID'),
                'Listings'  # Use the Listings table
            )
            
            # Format the listing data for Airtable
            fields = {
                'id': listing_data['listing_id'],
                'swarm_id': listing_data['swarm_id'],
                'number_of_shares': listing_data['number_of_shares'],
                'price_per_share': listing_data['price_per_share'],
                'total_price': listing_data['total_price'],
                'seller_wallet': listing_data['seller'],
                'listing_date': datetime.now().isoformat(),
                'status': 'active',
                'token_type': listing_data['token']
            }
            
            # Create the record
            record = listings_table.create(fields)
            return record
            
        except Exception as e:
            print(f"Error creating listing in Airtable: {e}")
            raise e

File: utils/airtable.js
class AirtableClient {
  constructor() {
    this.apiKey = process.env.AIRTABLE_API_KEY;
    this.baseId = process.env.AIRTABLE_BASE_ID;
    this.tableName = process.env.AIRTABLE_TABLE_NAME;
  }

  async update_user_status(telegram_id, status) {
    try {
      const response = await fetch(`https://api.airtable.com/v0/${this.baseId}/${this.tableName}`, {
        method: 'GET',
        headers: {
          'Authorization': `Bearer ${this.apiKey}`,
          'Content-Type': 'application/json'
        }
      });

      const data = await response.json();
      const user = data.records.find(record => record.fields.telegram_id === telegram_id);

      if (user) {
        await fetch(`https://api.airtable.com/v0/${this.baseId}/${this.tableName}/${user.id}`, {
          method: 'PATCH',
          headers: {
            'Authorization': `Bearer ${this.apiKey}`,
            'Content-Type': 'application/json'
          },
          body: JSON.stringify({
            fields: {
              status: status,
              subscription_updated: new Date().toISOString()
            }
          })
        });
        return true;
      }
      return false;
    } catch (error) {
      console.error('Airtable update error:', error);
      return false;
    }
  }

  async update_payment_details(telegram_id, details) {
    try {
      const response = await fetch(`https://api.airtable.com/v0/${this.baseId}/${this.tableName}`, {
        method: 'GET',
        headers: {
          'Authorization': `Bearer ${this.apiKey}`,
          'Content-Type': 'application/json'
        }
      });

      const data = await response.json();
      const user = data.records.find(record => record.fields.telegram_id === telegram_id);

      if (user) {
        await fetch(`https://api.airtable.com/v0/${this.baseId}/${this.tableName}/${user.id}`, {
          method: 'PATCH',
          headers: {
            'Authorization': `Bearer ${this.apiKey}`,
            'Content-Type': 'application/json'
          },
          body: JSON.stringify({
            fields: {
              payment_date: details.payment_date,
              payment_amount: details.amount_paid,
              transaction_id: details.transaction_signature,
              payment_status: details.payment_status
            }
          })
        });
        return true;
      }
      return false;
    } catch (error) {
      console.error('Airtable payment update error:', error);
      return false;
    }
  }

  async createListing(listingData) {
    try {
        // Get token label safely with fallback
        const tokenLabel = listingData.token ? listingData.token.label : 'USDC';

        const response = await fetch(`https://api.airtable.com/v0/${this.baseId}/Listings`, {
            method: 'POST',
            headers: {
                'Authorization': `Bearer ${this.apiKey}`,
                'Content-Type': 'application/json'
            },
            body: JSON.stringify({
                records: [{
                    fields: {
                        id: listingData.listing_id || `L${Math.floor(100000 + Math.random() * 900000)}`,
                        swarm_id: listingData.swarm_id,
                        number_of_shares: listingData.number_of_shares,
                        price_per_share: listingData.price_per_share,
                        total_price: listingData.total_price,
                        seller_wallet: listingData.seller,
                        listing_date: new Date().toISOString(),
                        status: 'active',
                        token_type: tokenLabel
                    }
                }]
            })
        });

        const data = await response.json();
        
        if (!data.records || data.records.length === 0) {
            throw new Error('Failed to create listing record');
        }

        console.log('Created listing:', data.records[0]);
        return data.records[0];
    } catch (error) {
        console.error('Airtable listing creation error:', error, listingData);
        throw error;
    }
  }
}

export { AirtableClient };
